### ISSUE TITLE
extract the feature of conv2 or fc1

### ISSUE THREAD

**Issue by stellaxu** [ray-project/ray#3660]
Hi, based on the visionnet model in rllib, the structure of DNN is: input->conv1->conv2->fc1->fc2.

My question is that how to modify the codes in Rllib to extract the features on conv2 layer or fc1 layer? I have tried several ways but cannot pass the compiler. Thanks.

For ES codes, only the feature on fc2 can be extracted by the following two parts:
1) From rllib/models/visionnet.py
class VisionNetwork(Model):

    def _build_layers_v2(self, input_dict, num_outputs, options):
        inputs = input_dict["obs"]
        filters = options.get("conv_filters")
        if not filters:
            filters = get_filter_config(options)

        activation = get_activation_fn(options.get("conv_activation", "relu"))

        with tf.name_scope("vision_net"):
            for i, (out_size, kernel, stride) in enumerate(filters[:-1], 1):
                inputs = slim.conv2d(
                    inputs,
                    out_size,
                    kernel,
                    stride,
                    activation_fn=activation, 
                    scope="conv{}".format(i))

            out_size, kernel, stride = filters[-1]
           #out_size = filters[-1,0]
            inputs = slim.flatten (inputs, scope="embed")
            fc1 = slim.fully_connected(inputs, out_size, activation_fn=activation, scope="fc1")
            fc2 = slim.fully_connected(fc1, num_outputs, activation_fn=None, normalizer_fn=None,  scope="fc2")

            return fc2, fc1

2) From rllib/agents/es/policies.py:
class GenericPolicy(object):
    def __init__(self, sess, env,env2,action_space, obs_space, preprocessor,
                 observation_filter, model_options, action_noise_std):
        self.sess = sess
        self.action_space = action_space
        self.action_noise_std = action_noise_std
        self.preprocessor = preprocessor
        self.observation_filter = get_filter(observation_filter,
                                             self.preprocessor.shape)
        self.inputs = tf.placeholder(tf.float32,
                                     [None] + list(self.preprocessor.shape))

        # Policy network.
        dist_class, dist_dim = ModelCatalog.get_action_dist(
            self.action_space, model_options, dist_type="deterministic")

        self.model = ModelCatalog.get_model({
            "obs": self.inputs}, obs_space, dist_dim, model_options)

        self.dist = dist_class(self.model.outputs)
        self.sampler = self.dist.sample()
        self.entro=self.dist.entropy()
        self.prob=self.dist.softmax()

        self.variables = ray.experimental.TensorFlowVariables(
            self.model.outputs, self.sess)

        self.num_params = sum(
            np.prod(variable.shape.as_list())
            for _, variable in self.variables.variables.items())

How to modify part 1) or part 2) to extract the feature layer of conv2 or fc1 especially for ES?

Thanks a lot.

**Comment by ericl** [comment]
If you use a custom model, then you can return whatever tensor you want as the feature layer and access it via model.last_layer (or define it as an extra attribute on the model).

**Comment by stellaxu** [comment]
Thanks Eric, I got it~~
